src/samVisualize/commands/visualizeSamTemplate/samVisualization.ts

```diff
@@ -0,0 +1,227 @@
+                        const pos = this.resourceLineMap![message.data]
+                        if (pos) {
+                            const decorations: vscode.DecorationOptions[] = []
+                            const startPosition = activeEditor.document.positionAt(pos.start).with({ character: 0 })
+                            const endPosition = activeEditor.document.positionAt(pos.end)
+                            const range = new vscode.Range(startPosition, endPosition)
+                            activeEditor.revealRange(range, vscode.TextEditorRevealType.InCenter)
+                            decorations.push({ range })
+                            activeEditor.setDecorations(decorationType, decorations)
+                        }
+                    })
+            })
+        )
+
+        panel.onDidDispose(() => {
+            this.webviewPanel = undefined
+            debouncedUpdate.cancel()
+            decorationType.dispose()
+            this.onVisualizationDisposeEmitter.fire()
+            this.disposables.forEach(disposable => {
+                disposable.dispose()
+            })
+            this.onVisualizationDisposeEmitter.dispose()
+        })
+
+        return panel
+    }
+
+    // Updates the visualization with the newly edited YAML
+    private updateVisualization(yamlString: string) {
+        const newGraphObject = generateGraphFromYaml(yamlString)
+
+        this.webviewPanel?.webview.postMessage({
+            graphObject: newGraphObject,
+        })
+
+        // Only generate a ResourceLineMap if the input has generated a valid GraphObject.
+        if (newGraphObject) {
+            const newResourceLineMap = generateResourceLineMap(yamlString)
+            // Update the resourceLineMap
+            this.resourceLineMap = newResourceLineMap
+        }
+    }
+
+    private getWebviewContent(
+        graphObjectString: string,
+        primaryResourceListString: string,
+        iconDataString: string,
+        webviewStylesheetUri: vscode.Uri,
+        webviewRenderScriptUri: vscode.Uri,
+        cspSource: string
+    ) {
+        // Library exported by webpack is GraphRender
+        return `<!DOCTYPE html>
+    <html lang="en">
+    <head>
+        <meta charset="UTF-8">
+        <meta http-equiv="Content-Security-Policy" 
+        content="default-src 'none'; 
+        img-src ${cspSource} https:; 
+        script-src ${cspSource} 'self' 'unsafe-inline' 'unsafe-eval'; 
+        style-src ${cspSource}; "/>
+        <meta name="viewport" content="width=device-width, initial-scale=1.0">
+        <link rel="stylesheet" href='${webviewStylesheetUri}'>
+        <script src='${webviewRenderScriptUri}'></script>
+        <title>Template Graph</title>
+
+    </head>
+    <body>
+        <script>
+            new GraphRender.ForceDirectedGraph(${graphObjectString},${primaryResourceListString},${iconDataString})
+        </script>
+    </body>
+    </html>`
```

Comment:
Is there any other way to inject JS code into a web content without passing them as strings?

Also wonder is there a way to make this html template an external file?

Reply:
There are ways, though I think this is up to the Toolkit to figure out which is the best way. We have a lot of disorganization in regards to how webviews are currently implemented.